op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+I.e. create an event for a deriver to turn attributes in an unsafe block,
+and follow that up with events to try to promote the block safety, suggest it for finality, and persisting it to the safe head DB.
+rather than tightly coupling the four acts: (1) block processing, (2) updating of the safe-head, (3) remembering the finality data all at once, (4) persisting the safe head change.
+
+
+#### Threads and parallel derivers
+
+Stacks can define a single `OnEvent` handler, and by using `Resource`, still execute on parallel threads, if desired.
+Note that things can be entirely synchronous, if the caller ignores the resource preference, and just runs one thing at a time. This can support fault-proofs, reducing the custom code (no fault-proof specific driver logic, just a synchronous loop of calling the deriver of choice).
+
+To get the parallelism, the `Driver`, which itself can be the absolute root `Deriver`,
+can create a new go routine for each new `Resource` type, when first seen.
+```go
+worker, ok := driver.workers[event.Resource]
+if !ok {
+    worker = ... // spawn new worker
+    driver.workers[event.Resource] = worker // keep it around
+}
+worker.OnEvent(ctx, event)
+```
+
+#### Instrumentation
+
+A `InstrumentedDeriver` can be introduced, as a shim for `OnEvent`, wrapping an underlying `Deriver`,
+to provide a default set of instrumentation for all current future event handling:
+- Measure the execution-time per event type.
+- Track utilization per resource.
+- Trace/debug log the results and timeouts.
+
+This can help identify bottlenecks, does not have to be rewritten with each op-stack mod/extension,
+and fits it all in just a few reusable/aggregate metrics charts.
+
+### Spec changes
+
+In the specs we enshrine:
+- The abstract idea of derivers as modules that can handle events.
+- The standard derivers and event-types for rollups.
+- For each standard deriver: the standard conditions, and corresponding effects, to perform.
+
+Alt-DA, interop, and other more experimental features can then introduce specs for additional derivers.
+
+We can do this incrementally:
+1. Introduce derivers as concept
+2. Document existing triggers/conditions/effects, grouped by (sub)deriver. (no breaking changes!)
+3. Introduce new derivers for Alt-DA and Interop etc.
+
+Conditions/Effects to specify (grouped by resource here) loosely taken from [specs PR 96](https://github.com/ethereum-optimism/specs/pull/96):
+
+#### Unsafe head
+
+Triggers:
+- Change of unsafe-payloads queue
+- Changed unsafe L2 head
+- Time to sequence (open or seal block)
+- Attributes conflicting with chain
+- Attributes applicable to head
+- (new, maybe) Unsafe-head L1-origin not in or ahead of canonical L1
+- (new, maybe) Sync-status change (if not implied by unsafe L2 head change)
+
+Condition/effect pairs:
+- Unsafe block addition
+- Unsafe block sync-trigger
+- Unsafe block processing
+- Sequencing
+- Payload attributes processing
+
+#### Cross-unsafe head (interop)
```

Comment:
Could you elaborate on the example of the cross-unsafe head how this new head stage could be added in a modular way without touching the resource implementations of the unsafe and safe head? IIRC the safe head would be dependent on the cross-unsafe in that it would be a new dependent intermediary head stage.

Reply:
Two main reusable building blocks:
- The safe-head handling can respond to events that describe when something can be promoted promoted.
- The unsafe-head handling can take responsibility over processing block-attributes into new blocks, and emit events that signal the unsafe-head change with the context of the block-attributes causing that change.

We can compose the above with one of two options:
- Regular: Translate the unsafe-block-from-attributes signal to a promote-to-safe signal
- Interop: Accept unsafe-block-from-attributes as trigger to check cross-unsafe status.

So depending on regular/interop mode, we can just translate one event in the other, to trigger the right behavior. Or in other words, if we don't tie together the exact derivers events, we can compose them by translating the events any way we like.